cmd/generic-handlers.go

```diff
@@ -311,21 +305,41 @@ func hasBadHost(host string) error {
 // Check if the incoming path has bad path components,
 // such as ".." and "."
 func hasBadPathComponent(path string) bool {
-	if len(path) > 4096 {
-		// path cannot be greater than Linux PATH_MAX
-		// this is to avoid a busy loop, that can happen
-		// if the caller sends path of following style
-		// a/a/a/a/a/a/a/a...
+	n := len(path)
+	if n > 32<<10 {
+		// At 32K we are beyond reasonable.
 		return true
 	}
-	path = filepath.ToSlash(strings.TrimSpace(path)) // For windows '\' must be converted to '/'
-	for _, p := range strings.Split(path, SlashSeparator) {
-		switch strings.TrimSpace(p) {
-		case dotdotComponent:
+	i := 0
+	// Skip leading slashes (for sake of Windows \ is included as well)
+	for i < n && (path[i] == SlashSeparatorChar || path[i] == '\\') {
+		i++
+	}
+
+	for i < n {
+		// Find the next segment
+		start := i
+		for i < n && path[i] != SlashSeparatorChar && path[i] != '\\' {
+			i++
+		}
+
+		// Trim whitespace of segment
```

Comment:
@klauspost Was the trimming of whitespace a new addition in this PR? https://github.com/minio/minio/issues/21022 reports that S3 considers whitespace before and after `.` to be valid:
```
$ echo hello | aws --profile jrobinso-aws s3 cp - s3://test-jrobinso/\.\ /hello
$ aws --profile jrobinso-aws s3 ls --recursive s3://test-jrobinso/
2025-03-12 11:04:55          6 . /hello
```

```
$ echo hello | aws --profile test-elm s3 cp - s3://test-jrobinso/\.\ /hello
upload failed: - to s3://test-jrobinso/. /hello An error occurred (XMinioInvalidResourceName) when calling the PutObject operation: Resource name contains bad components such as ".." or ".".
```

Reply:
Old code had `path = filepath.ToSlash(strings.TrimSpace(path))`.

Reply:
I don't really care about the case in the linked PR. AWS also says to not use silly object names.

Reply:
@klauspost  it came from an actual research user data at Stanford, likely a program creating those weird `. ` path components, that work with AWS, but not when we tested migrating them to MinIO. Just wanted to emphasis that while it looks silly, it is based on actual user data and supported by AWS.